Alex and Happy Part 1/article.md

```diff
@@ -0,0 +1,817 @@
+
+For now, at the top of our file, we only declare the module name, but we will soon increase our export list and add more imports.
+
+In the middle section, we first declare our _wrapper_, which indicates the type of code Alex should generate for us (`monadUserState`, which allows us to save custom state) and the input type (`bytestring`, but we could use plain Haskell `String`s instead).
+Consult the Alex User Manual on [Wrappers](https://haskell-alex.readthedocs.io/en/latest/api.html#wrappers) if you want to see other wrappers.
+
+The following section is `tokens :-`, where we will list all the patterns defined by regular expressions to match lexemes in our grammar plus an action on what the lexer should do with the matched lexeme.
+The first definition we provided is `<0> $white+ ;`, which simply indicates that all white space should be skipped.
+More on this later.
+
+The bottom section contains some boilerplate that Alex requires us to write. These include a data type that needs to be called `AlexUserState`, a value with the initial state called `alexInitUserState`, and a value called `alexEOF`, which instructs Alex how to build the EOF (End-Of-File) token, reached when Alex has finished lexing the input string.
+
+We add some additional datatypes: `Range`, `RangedToken`, and `Token`, which we will use throughout the article to describe the tokens that we've successfully created, as well as their positions. Saving the ranges is unnecessary, but it's an excellent addition when reporting errors.
+
+Finally, for the EOF token, we use the `alexGetInput` action to retrieve the current position of the scanner and provide it to the token.
+
+We should provide one token name for each lexical category in the grammar.
+You are free to add new token names as you wish. Still, for this tutorial, we will have the following: operators, keywords, literals (strings and integers), identifiers, a token representing the end-of-file (EOF), etc.
+
+Our token datatype is thus:
+
+```haskell
+data Token
+  -- Identifiers
+  = Identifier ByteString
+  -- Constants
+  | String ByteString
+  | Integer Integer
+  -- Keywords
+  | Let
+  | In
+  | If
+  | Then
+  | Else
+  -- Arithmetic operators
+  | Plus
+  | Minus
+  | Times
+  | Divide
+  -- Comparison operators
+  | Eq
+  | Neq
+  | Lt
+  | Le
+  | Gt
+  | Ge
+  -- Logical operators
+  | And
+  | Or
+  -- Parenthesis
+  | LPar
+  | RPar
+  -- Lists
+  | Comma
+  | LBrack
+  | RBrack
+  -- Types
+  | Colon
+  | Arrow
+  -- EOF
+  | EOF
+  deriving (Eq, Show)
+```
+
+At the top of the file, make sure to include the appropriate imports:
+
+```haskell
+import Data.ByteString.Lazy.Char8 (ByteString)
+import qualified Data.ByteString.Lazy.Char8 as BS
+```
+
+### Regex and regex macros
```

Comment:
There seems to be a bit of a chaotic moving here from regex to identifier to regex to identifier. 

What I would suggest is to call this section Lexing identifiers, and put the text about regex macros in a block like this.
```
<hr>
**Regex macros**

text
<hr>
```
Here's an example of how that looks: https://serokell.io/blog/untyped-lambda-calculus#defining-computation